Migrate from `Virtus` to `ActiveModel::Attributes` (and get rid of `Rectify::Form`)

ahukkanen opened the pull request:
#### :tophat: What? Why?
The basic idea of this PR is to refactor the Virtus models (such as all instances of `Rectify::Form`) in the core to the new ActiveModel::Attributes based solution as discussed at #7234. The reason for this is the discontinuation of Virtus.

This provides a compatibility layer that preserves backwards compatibility as long as possible with the Rectify::Forms API requiring little to no changes in the existing Form classes. ActiveModel::Attributes works a bit differently than Rectify::Form (e.g. no nested Forms or attributes), so we need this layer to keep functionality as-is without any major changes.

#### :pushpin: Related Issues
*Link your PR to an issue*
- Related to #7234
- Fixes #8514

#### Testing
Run the specs and test the application with this implementation.

#### :clipboard: Checklist

- [x] :question: **CONSIDER** adding a unit test if your PR resolves an issue.
- [x] :heavy_check_mark: **DO** check open PR's to avoid duplicates.
- [x] :heavy_check_mark: **DO** keep pull requests small so they can be easily reviewed.
- [x] :heavy_check_mark: **DO** build locally before pushing.
- [x] :heavy_check_mark: **DO** make sure tests pass.
- [ ] :heavy_check_mark: **DO** make sure any new changes are documented in `docs/`.
- [ ] :heavy_check_mark: **DO** add and modify seeds if necessary.
- [x] :heavy_check_mark: **DO** add CHANGELOG upgrade notes if required.
- [ ] :heavy_check_mark: **DO** add to GraphQL API if there are new public fields.
- [ ] :heavy_check_mark: **DO** add link to MetaDecidim if it's a new feature.
- [x] :x:**AVOID** breaking the continuous integration build.
- [ ] :x:**AVOID** making significant changes to the overall architecture.

ahukkanen commented:
@andreslucena This is now (finally) ready for review.

Sorry for the slight delay, the scope was a bit more than expected (as usual with the bigger changes). I wanted to make sure this is done well as it's pretty integral part of the Decidim core.

andreslucena reviewed (changes requested):
Great job!! This is a much-needed change 👍🏽 
I have a couple typos to correct, a couple doubts to solve, and also I think that we should document a bit better with top-level comments the new `Decidim::AttributeObject` classes.

andreslucena commented:
Also, there are merge conflicts. It's almost sure the CHANGELOG, as I freeze features for v0.26 earlier this week, so we have more time to try this and other possible breaking changes in `develop`. Can you solve the conflict please?

ahukkanen commented on decidim-core/lib/decidim/attribute_object.rb:
```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
+
+module Decidim
+  module AttributeObject
```
Fixed by a4fa021c4aa6ca560790ae89619f2c8f4a37262a.

ahukkanen commented on decidim-core/lib/decidim/attribute_object/form.rb:
```diff
@@ -0,0 +1,158 @@
+
+module Decidim
+  module AttributeObject
+    class Form
```
Fixed by a4fa021c4aa6ca560790ae89619f2c8f4a37262a.

ahukkanen commented on decidim-core/lib/decidim/attribute_object/form.rb:
```diff
@@ -0,0 +1,158 @@
+        class_name.chomp("Form").underscore.to_sym
+      end
+
+      # Converts the mimiced name to ActiveModel naming.
```
Fixed by b29f4d51370b2769ec477beb738f020be5bece49.

ahukkanen commented on decidim-conferences/spec/forms/registration_type_form_spec.rb:
```diff
@@ -91,7 +91,9 @@ module Admin
           let(:price) { "blabla" }
 
-          it { is_expected.to be_invalid }
+          # The value is cast to the correct type so "blabla" becomes 0.0 which
```
Yes, exactly this is the reason. Same thing for `ActiveModel::Type::Integer.new.cast("blabla")`.

I don't think it's that bad actually as most of the times those inputs should be `<input type="number">` which shouldn't allow adding other inputs than numbers. Firefox is an exception to that rule but I don't think it's that big of an issue in actual use cases that `blabla` (or any string) would be converted to zeros as that input should be disabled by the UI in the ideal situation (such as Chrome).

Empty string input will still cast to `nil` which allows adding empty values to these fields.

andreslucena replied:
Ok, fair enough, for me the only problem is how the spec is read:

```ruby
let(:price) { "blabla" }
it { is_expected.to be_valid }
```
But with your comment, it is clear that this is a weird behavior indeed.

Also, I've checked it and for end users there isn't any behavior change, when adding a string in this field in Firefox it saves it with as an empty string on both cases (`Virtus` and `Attributes` API), so that's fine for me.

ahukkanen commented on CHANGELOG.md:
```diff
@@ -28,6 +28,25 @@ PR [\#8658](https://github.com/decidim/decidim/pull/8658) Blocked users are pres
+- Very rarely, when defining a an attribute of type `Rails::Engine`, you need to change `attribute :attr_name, Rails::Engine` to `attribute :attr_name, Rails::Engine, {}`. This is because we want to preserve the method signature against `ActiveModel::Attributes` for the `attribute` class method intead of the legacy `Virtus.model`. There is a limitation in the Ruby language that if the method has  ...
+- If you had overridden any of the [`Rectify::Form` methods](https://github.com/andypike/rectify/blob/v0.13.0/lib/rectify/form.rb) within your form classes, remove those overrides. For example, you might have overridden the `form_attributes_valid?` method which no longer does anything. Instead, define a custom validation in order to add extra validations to your forms.
+- Test all your form and command classes thoroughly to notice any differences between the two implementations. The new layer is a bit more "robust" with some of the type castings, so some things may break during the migration in case you have relied on some of the oversights within `Virtus`.
+
```
Definitely we should and I forgot about that (it was on my mind before).

Fixed by ba3ffc7d6aaed06c55192f3342baefc3c04f0f86.

ahukkanen commented on CHANGELOG.md:
```diff
@@ -28,6 +28,25 @@ PR [\#8658](https://github.com/decidim/decidim/pull/8658) Blocked users are pres
+- Change all `attribute :attr_name, Hash` to `attribute :attr_name, Hash[Symbol => ExpectedType]` where `ExpectedType` is the type you are expecting the hash values to be. The new layer will default the hash key types to `Symbol` and hash value types to `Object` (= any type). The Virtus Hash attribute did not force any default types for these. It should be preferred to use the actual expected typ ...
+- Change all `attribute :attr_name, Array` to `attribute :attr_name, Array[ExpectedType]` where `ExpectedType` is the type you are expecting the array values to be. It should be preferred to use the actual expected type for the values instead of `Object` (= any type) to make your code more robutst and less buggy.
+- The original form attribute values are no longer available through the `@attr_name` instance variables within the Form or `Virtus.model` classes. Instead, change all these references to `@attributes["attr_name"].value` in case you want to fetch the original value of the attribute without using its accessor method. Another way is to provide an alias for the original attribute method before overr ...
+- When calling the `attributes` method of the model/form classes, use strings to refer to the attribute names, not symbols as you might have done with `Virtus` or `Rectify::Form`. Change all `model.attributes[:attr_name]` method calls to `model.attributes["attr_name"]`.
```
This is something I was also thinking before. Some versions in the implementation history did just that but I feel that when we are migrating to `ActiveModel::Attributes`, we should respect their APIs instead of the Virtus APIs.

By default `ActiveModel::Attributes` only responds to the strings hash keys.

We could "fix" that by defining an override to the `attributes` method as `def attributes ; super.with_indifferent_access ; end` but I would see this quite hacky solution as going further I think we should match the API that is provided by `ActiveModel::Attributes`. It might also have some small performance impacts.

I was also thinking of adding a migration configuration (disabled by default) which would conditionally add this explained override to the `attributes` method but I am strongly adhering towards the API we are relying on under the middle-layer.

This is also why I mentioned in the CHANGELOG comments that this "aims to provide as much backwards compatibility as possible". This is one place where I would see the need for change honestly, although I completely understand your point of view.

In real life I think most of the times the required changes should be in specs, as in the forms, commands and controllers, the programmers should always prefer to call the accessor methods.

alecslupu replied:
@ahukkanen, i would go for 1-2 releases with the override mention, and a deprecation warning. 

```
def attributes 
  super.with_indifferent_access
end
```
This PR may kill most of the community based contributions like : 
- [Decidim Awesome](https://github.com/Platoniq/decidim-module-decidim_awesome)
- [NavBar Links](https://github.com/OpenSourcePolitics/decidim-module-navbar_links)

ahukkanen replied:
@alecslupu I have made this change. This will mean that 3rd party developers will not update their code and will rely on the symbol accessors as long as it works fine, although it is not compatible with the underlying API.

Regarding the community based 3rd party modules, could you specify more clearly what you are expecting to break in these modules so that I can test it? By quickly glancing I wouldn't expect anything to break in the NavbarLinks module and in Decidim Awesome I think most of the things will work just fine without changes.

As mentioned before, mostly the changes are required in the specs of the 3rd party modules (change `attributes` hash expectation keys from symbols to strings). Other than that, it should be mostly backwards compatible now even with the symbol accessors after the latest change.

Other way you could fix the breaking specs is to change `expect(obj.attributes).to eq(foo: "bar")` into `expect(obj.to_h).to eq(foo: "bar")`. The `to_h` method will return a hash with symbol keys.

alecslupu replied:
@ahukkanen, i will give it a spin and i would let you know if my assumptions are validated.

andreslucena replied:
The current solution is great for me. I'll add an issue to remember to remove this hack in v0.30.

ahukkanen commented on CHANGELOG.md:
```diff
@@ -28,6 +28,25 @@ PR [\#8658](https://github.com/decidim/decidim/pull/8658) Blocked users are pres
+
+- Change all instances of `YourForm < Rectify::Form` to `YourForm < Decidim::Form`. It should be very rare to find any classes in your code that inherit directly from `Rectify::Form` but in case you have used that, replace those references with `Decidim::Form`.
+- Change all instances `include Virtus.model` to `include Decidim::AttributeObject::Model`.
+- Change all `attribute :attr_name, Hash` to `attribute :attr_name, Hash[Symbol => ExpectedType]` where `ExpectedType` is the type you are expecting the hash values to be. The new layer will default the hash key types to `Symbol` and hash value types to `Object` (= any type). The Virtus Hash attribute did not force any default types for these. It should be preferred to use the actual expected typ ...
```
Fixed by ae7ef9a95a87a990ed8d82d6da77629714ce2dff.

ahukkanen commented:
@andreslucena Comments resolved (that I could resolve) and merge conflicts resolved regarding the CHANGELOG.

andreslucena reviewed (approved):
All is OK for me! 👍🏽

andreslucena commented:
Can we merge this one @ahukkanen? Or is there anything else that you want to add to this PR?

ahukkanen commented:
@andreslucena On my part it's ready.

@alecslupu Raised [some concerns](https://github.com/decidim/decidim/pull/8669#discussion_r785683085) regarding compatibility with 3rd party modules. I thought I'd still test those two modules he mentioned.

I'm not expecting zero changes in these but I would expect most of the things to work fine without changes. Specs will likely need some updates, though.

ahukkanen commented:
@alecslupu @andreslucena 

I have tested both NavbarLinks and Decidim Awesome modules with the `ActiveModel::Attributes` changes.

Here's my findings:

## NavbarLinks

Branch `master`, changed the version number to `~> 0.27.0.dev`.

No problems identified. No changes done.

All other specs pass except this one:
https://github.com/OpenSourcePolitics/decidim-module-navbar_links/blob/6de93c18ebaa3a788b1e1ba26f1c771cc319aa51/spec/system/navbar_links/navbar_links_spec.rb#L16

This failing spec is **not** related to the `ActiveModel::Attributes` migration, it is failing because of the external link warnings added at #7397 which causes the unrecognized external URLs to be linked through `/link?external_url=http...`.

## Decidim Awesome

Branch `upgrade_025`, changed the version number to `~> 0.27.0.dev`. Also had to add admin engines to the Awesome modules (iframe and map components) for them not to throw any exceptions in the admin panel participatory space views. For the specs to run had to change factory `:editor_image` to `:awesome_editor_image` because this factory now already exists in the core.

Working fine:
- Dropping image to a WYSIWYG field (image saved as a blob)
- Updating constraints for configs

Not working without changes:
- Autosave for surveys (but I think this is some other problem since it should work in the front-end, right?)
- Updating awesome configs -> mismatch for string keys in the attributes https://github.com/Platoniq/decidim-module-decidim_awesome/blob/8b36e35a22ca22d7ff21ef4bf6ed9afcfa47cb98/app/forms/decidim/decidim_awesome/admin/config_form.rb#L34 -- Fix: `key.to_sym` -> `key.to_s`

After the above fix, all configuration forms work normally in the admin panel.

All other non-system specs pass after this change except for the following specs:
- `./spec/uploaders/image_uploader_spec.rb` - Needs the module to be updated to support Webpacker
- `./spec/lib/system_checker_spec.rb` - Something with the checksums/decidim version changes did not check further
- `./spec/forms/admin/menu_form_spec.rb` - Label keys need to be converted from symbols to hashes in the expectation
- `./spec/commands/admin/update_constraint_spec.rb` - Extra `id` field in the constraint form attributes (spec update needed)
- `./spec/controllers/editor_images_controller_spec.rb` - Needs the module to be updated to support Webpacker
- `./spec/controllers/admin/checks_controller_spec.rb` - I believe needs some changes with the module versions (?)

I believe most of the system specs failures are caused by the missing assets (or broken front-end), as the module has not yet been updated to be compatible with Webpacker. It would make much more sense to test these after the module has been upgraded to 0.25 or 0.26.

## Further notes

I am no expert in either of these modules, these were just quick tests within the UI and by running the specs.

alecslupu commented:
> This failing spec is **not** related to the `ActiveModel::Attributes` migration, it is failing because of the external link warnings added at #7397 which causes the unrecognized external URLs to be linked through `/link?external_url=http...`.

Addressed here ... pending @armandfardeau 's input : https://github.com/OpenSourcePolitics/decidim-module-navbar_links/pull/23
